lib/JSON/Validator.pm

```diff
@@ -121,58 +110,38 @@ sub _build_formats {
 }
 
 sub _load_schema {
-  my ($self, $url, $parent) = @_;
-  my ($namespace, $scheme) = ("$url", "file");
-  $namespace =~ s|\\|/|g if $^O eq 'MSWin32';
-  my $doc;
-
-  if ($namespace =~ $HTTP_SCHEME_RE) {
-    $url = Mojo::URL->new($url);
-    ($namespace, $scheme) = ($url->clone->fragment(undef)->to_string, $url->scheme);
-  }
-  elsif ($namespace =~ m!^data://(.*)!) {
-    $scheme = 'data';
-  }
-  elsif ($parent and $parent =~ $HTTP_SCHEME_RE) {
-    $parent = Mojo::URL->new($parent);
-    $url =~ s!#.*!!;
-    $url = $parent->path($parent->path->merge($url)->canonicalize);
-    ($namespace, $scheme) = ($url->to_string, $url->scheme);
+  my ($self, $url) = @_;
+
+  if ($url =~ m!^https?://!) {
+    warn "[JSON::Validator] Loading schema from URL $url\n" if DEBUG;
+    return $self->_load_schema_from_url(Mojo::URL->new($url)->fragment(undef)), "$url";
   }
-  elsif ($parent) {
-    $url =~ s!#.*!!;
-    $url = path(path($parent)->dirname, split '/', $url);
-    $namespace = Cwd::abs_path($url->to_string) || $url;
+
+  if ($url =~ m!^data://([^/]+)/(.*)!) {
+    my ($module, $file) = ($1, $2);
+    warn "[JSON::Validator] Loading schema from data section: $url\n" if DEBUG;
+    my $text = Mojo::Loader::data_section($module, $file)
+      || Carp::confess("$file could not be found in __DATA__ section of $module.");
+    return $self->_load_schema_from_text(\$text), "$url";
   }
-  elsif ($url =~ m!^/! and !-e $url and $self->ua->server->app) {
-    $scheme = 'http';
+
+  if ($url =~ m!^\s*[\[\{]!) {
+    warn "[JSON::Validator] Loading schema from string.\n" if DEBUG;
+    return $self->_load_schema_from_text(\$url), '';
   }
 
-  # Make sure we create the correct namespace if not already done by Mojo::URL
-  $namespace =~ s!#.*$!! if $namespace eq $url;
-
-  return $doc if $doc = $self->_store(cached => $namespace);
-  return eval {
-    warn "[JSON::Validator] Loading schema $url namespace=$namespace scheme=$scheme\n"
-      if DEBUG;
-    $doc
-      = $scheme eq 'file' ? path($namespace)->slurp
-      : $scheme eq 'data' ? $self->_load_schema_from_data($url, $namespace)
-      :                     $self->_load_schema_from_url($url, $namespace);
-    $self->_register_document($self->_load_schema_from_text(\$doc), $namespace);
-  } || do {
-    $doc ||= '';
-    die "Could not load document from $url: $@ ($doc)" if DEBUG;
-    die "Could not load document from $url: $@";
-  };
-}
+  my $file = $url;
+  $file =~ s!^file://!!;
+  $file =~ s!#$!!;
+  $file = path(split '/', $file);
+  if (-e $file) {
+    $file = $file->realpath;
+    $url = sprintf "file://%s", join "/", @$file;
```

Comment:
This is not a good idea. Need to figure out some way to keep relative URLs as-is, so external document renderers download the relative `$ref`.